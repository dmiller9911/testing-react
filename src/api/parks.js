export function getParks() {
  return fetch('/parks.json').then(res => res.json());
}
