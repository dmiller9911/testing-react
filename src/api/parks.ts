export interface Park {
  name: string;
  location: string;
  maxElevation: number;
  minElevation: number;
  size: number;
  establishedDate: string;
  id: string;
}

export function getParks(): Promise<Park[]> {
  return fetch('/parks.json').then(res => res.json());
}
