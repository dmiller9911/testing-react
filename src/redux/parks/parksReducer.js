import {
  PARKS_FETCH_ALL_FAIL,
  PARKS_FETCH_ALL_REQUEST,
  PARKS_FETCH_ALL_SUCCESS,
  PARKS_TOGGLE_FAVORITE,
  PARKS_FILTER,
  parkFilters
} from './parksActions';

const defaultState = {
  byId: {},
  favorites: new Set(),
  error: null,
  isLoading: false,
  filter: parkFilters.all
};

export const stateKey = 'parks';

export function parksReducer(state = defaultState, action) {
  switch (action.type) {
    case PARKS_FETCH_ALL_REQUEST:
      return {
        ...state,
        isLoading: true
      };

    case PARKS_FETCH_ALL_SUCCESS:
      return {
        ...state,
        byId: action.payload.reduce(
          (newState, park) => ({
            ...newState,
            [park.id]: park
          }),
          {}
        ),
        error: null,
        isLoading: false
      };

    case PARKS_FETCH_ALL_FAIL:
      return {
        ...state,
        isLoading: false,
        error: action.payload
      };

    case PARKS_TOGGLE_FAVORITE:
      const parkId = action.payload.id;
      const favorites = new Set(state.favorites);
      if (favorites.has(parkId)) {
        favorites.delete(parkId);
      } else {
        favorites.add(parkId);
      }
      return {
        ...state,
        favorites
      };

    case PARKS_FILTER:
      return {
        ...state,
        filter: action.payload
      };

    default:
      return state;
  }
}
