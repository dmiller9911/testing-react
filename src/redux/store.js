import {
  applyMiddleware,
  compose,
  createStore,
  DeepPartial,
  combineReducers
} from 'redux';
import thunk from 'redux-thunk';
import { rootReducer, RootState } from './rootReducer';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const middlewares = [thunk];
const enhancer = composeEnhancers(applyMiddleware(...middlewares));

export function configureStore(initialState) {
  return createStore(rootReducer, initialState, enhancer);
}

export function createMockStoreCreator(reducer) {
  return initialState => {
    return createStore(
      typeof reducer === 'function' ? reducer : combineReducers(reducer),
      initialState,
      applyMiddleware(...middlewares)
    );
  };
}
