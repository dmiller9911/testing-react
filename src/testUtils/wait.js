export function wait() {
  return new Promise(resolve => {
    setImmediate(resolve);
  });
}
