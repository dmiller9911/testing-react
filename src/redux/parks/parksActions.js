import { getParks } from '../../api/parks';

export const PARKS_FETCH_ALL_REQUEST = 'parks/fetch_all_request';
export const PARKS_FETCH_ALL_SUCCESS = 'parks/fetch_all_success';
export const PARKS_FETCH_ALL_FAIL = 'parks/fetch_all_fail';
export const PARKS_TOGGLE_FAVORITE = 'parks/toggle_favorite';
export const PARKS_FILTER = 'parks/filter';

export const parkFilters = {
  all: 'all',
  favorites: 'favorites'
};

export const getAllParksRequest = () => ({ type: PARKS_FETCH_ALL_REQUEST });
export const getAllParksSuccess = parks => ({
  type: PARKS_FETCH_ALL_SUCCESS,
  payload: parks
});
export const getAllParksFailure = err => ({
  type: PARKS_FETCH_ALL_FAIL,
  payload: err
});
export function getAllParks() {
  return dispatch => {
    dispatch(getAllParksRequest());
    getParks()
      .then(parks => dispatch(getAllParksSuccess(parks)))
      .catch(err => dispatch(getAllParksFailure(err)));
  };
}

export function toggleFavorite(park) {
  return {
    type: PARKS_TOGGLE_FAVORITE,
    payload: park
  };
}

export function setParkFilter(filter) {
  return {
    type: PARKS_FILTER,
    payload: filter
  };
}
