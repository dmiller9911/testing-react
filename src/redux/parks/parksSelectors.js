import { createSelector } from 'reselect';
import { parkFilters } from './parksActions';
import { stateKey } from './parksReducer';

export const selectParksState = state => state[stateKey];
export const selectParks = state => Object.values(selectParksState(state).byId);
export const selectParkFilter = state => selectParksState(state).filter;
export const selectFavorites = state => selectParksState(state).favorites;
export const selectParksLoading = state => selectParksState(state).isLoading;

export const selectFilteredParks = createSelector(
  [selectParkFilter, selectParks, selectFavorites],
  (filter, parks, favorites) => {
    switch (filter) {
      case parkFilters.favorites:
        return parks.filter(p => favorites.has(p.id));
      case parkFilters.all:
      default:
        return parks;
    }
  }
);
