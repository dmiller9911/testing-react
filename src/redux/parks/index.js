import * as parksActions from './parksActions';
import * as parksSelectors from './parksSelectors';
import { parksReducer, stateKey as parksStateKey } from './parksReducer';

export { parksReducer, parksActions, parksSelectors, parksStateKey };
