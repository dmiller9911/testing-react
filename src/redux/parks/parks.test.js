jest.mock('../../api/parks');

import { createMockStoreCreator } from '../store';
import { parksReducer, stateKey } from './parksReducer';
import * as actions from './parksActions';
import { wait } from '../../testUtils/wait';
import { getParks } from '../../api/parks';
import * as selectors from './parksSelectors';

const createParksStore = createMockStoreCreator({ [stateKey]: parksReducer });
const park1 = { id: 'park1' };
const park2 = { id: 'park2' };

getParks.mockReturnValue(Promise.resolve([park1, park2]));

it('handles fetching all parks', async () => {
  const store = createParksStore();
  store.dispatch(actions.getAllParks());
  expect(selectors.selectParksLoading(store.getState())).toBe(true);
  await wait();
  expect(selectors.selectParks(store.getState())).toMatchSnapshot('success');
});

it('toggles a park as a favorite', () => {
  const store = createParksStore();
  store.dispatch(actions.toggleFavorite(park1));
  expect(selectors.selectFavorites(store.getState())).toMatchSnapshot(
    'adds favorite'
  );
  store.dispatch(actions.toggleFavorite(park1));
  expect(selectors.selectFavorites(store.getState())).toMatchSnapshot(
    'removes favorite'
  );
});

it('sets filter', () => {
  const store = createParksStore();
  store.dispatch(actions.setParkFilter(actions.parkFilters.favorites));
  expect(selectors.selectParkFilter(store.getState())).toBe(
    actions.parkFilters.favorites
  );
});
