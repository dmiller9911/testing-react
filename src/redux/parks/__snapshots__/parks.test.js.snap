// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles fetching all parks: success 1`] = `
Array [
  Object {
    "id": "park1",
  },
  Object {
    "id": "park2",
  },
]
`;

exports[`toggles a park as a favorite: adds favorite 1`] = `
Set {
  "park1",
}
`;

exports[`toggles a park as a favorite: removes favorite 1`] = `Set {}`;
